batch/batch/server/database.py

```diff
@@ -1,159 +1,133 @@
-            async with conn.cursor() as cursor:
-                where_template, where_values = make_where_statement(where_items)
-                sql = f"DELETE FROM `{self.name}` WHERE {where_template}"
-                await cursor.execute(sql, where_values)
+        await super().__init__(config_file)
+
+        self.jobs = await JobsTable(self, os.environ.get('JOBS_TABLE', 'jobs'))
+        self.jobs_parents = await JobsParentsTable(self, os.environ.get('JOBS_PARENTS_TABLE', 'jobs-parents'))
+        self.batch = await BatchTable(self, os.environ.get('BATCH_TABLE', 'batch'))
+        self.batch_jobs = await BatchJobsTable(self, os.environ.get('BATCH_JOBS_TABLE', 'batch-jobs'))
+
+
+class JobsTable(Table):
+    async def __init__(self, db, name='jobs'):
+        schema = {'id': 'BIGINT NOT NULL AUTO_INCREMENT',
+                  'state': 'VARCHAR(40) NOT NULL',
+                  'exit_code': 'INT',
+                  'batch_id': 'BIGINT',
+                  'scratch_folder': 'VARCHAR(1000)',
+                  'pod_name': 'VARCHAR(1000)',
+                  'pvc': 'TEXT(65535)',
+                  'callback': 'TEXT(65535)',
+                  'task_idx': 'INT NOT NULL',
+                  'always_run': 'BOOLEAN',
+                  'time_created': 'TIMESTAMP DEFAULT CURRENT_TIMESTAMP',
+                  'time_ended': 'TIMESTAMP',
+                  'user': 'VARCHAR(1000)',
+                  'attributes': 'TEXT(65535)',
+                  'tasks': 'TEXT(65535)',
+                  'parent_ids': 'TEXT(65535)',
+                  'input_log_uri': 'VARCHAR(1000)',
+                  'main_log_uri': 'VARCHAR(1000)',
+                  'output_log_uri': 'VARCHAR(1000)'}
+
+        keys = ['id']
+
+        await super().__init__(db, name, schema, keys)
+
+    async def update_record(self, id, **items):
+        await super(JobsTable, self).update_record({'id': id}, items)
+
+    async def get_records(self, ids, fields=None):
+        assert isinstance(ids, list)
+        return await super(JobsTable, self).get_record({'id': id}, fields)
+
+    async def get_record(self, id, fields=None):
+        records = await self.get_records({'id': id}, fields)
+        assert len(records) == 1
+        return records[0]
+
+    async def has_record(self, id):
+        return await super(JobsTable, self).has_record({'id': id})
+
+    async def delete_record(self, id):
+        await super(JobsTable, self).delete_record({'id': id})
+
+    async def get_incomplete_parents(self, id):
+        parent_ids = await self.get_record(id, ['parent_ids'])
+        parent_ids = json.loads(parent_ids['parent_ids'])
+        parent_records = await self.get_records(parent_ids)
+        incomplete_parents = [pr['id'] for pr in parent_records.result()
+                              if pr['state'] == 'Created']
+        return incomplete_parents
+
+
+class JobsParentsTable(Table):
+    async def __init__(self, db, name='jobs-parents'):
+        schema = {'id': 'BIGINT',
```

Comment:
Just checking that we want more than 4 billion items in this table.

Reply:
also, `id` here is probably `NOT NULL AUTO_INCREMENT` right?

Reply:
No. This is a relation from the id in `jobs` to its parent id which is also in `jobs`. So should be the same type as the id in jobs

Reply:
Ah right, this is an associative table, I see

Reply:
If that's the case, a more idiomatic way of naming this would be 'batch_id'. Naming the first key id makes it seem like you are using that as your primary key.